processing/src/main/java/org/apache/druid/frame/allocation/HeapMemoryAllocator.java

```diff
@@ -53,7 +53,7 @@ public static HeapMemoryAllocator unlimited()
   @Override
   public Optional<ResourceHolder<WritableMemory>> allocate(final long size)
   {
-    if (bytesAllocated < capacity - size) {
+    if (size <= capacity - bytesAllocated) {
```

Comment:
hmm can we have some tests cases which check this ?

Reply:
In practice, `capacity = Long.MAX_VALUE`, therefore this edge condition won't trigger ever in a meaningful way. One thing I can do is break the abstraction and add `VisibleForTesting` to the constructor for the check.

Reply:
Added a few test cases for this class